src/FarBackwardsBeamPipes_geo.cpp

```diff
@@ -0,0 +1,120 @@
+      lengths.push_back( getAttrOrDefault<double>(pipe, _Unicode(length), 0) );
+      thetas.push_back( getAttrOrDefault<double>(pipe, _Unicode(theta), 0) );
+      rOuters1.push_back( getAttrOrDefault<double>(pipe, _Unicode(rout1), 0) );
+      rOuters2.push_back( getAttrOrDefault<double>(pipe, _Unicode(rout2), 0) );
+    }
+    else{ // Left blank in xml and to be calculated below
+      xCenters.push_back( 0 );
+      zCenters.push_back( 0 );
+      lengths.push_back( 0 );
+      thetas.push_back( 0 );
+      rOuters1.push_back( 0 );
+      rOuters2.push_back( 0 );
+    }
+  }
+
+  // Calculate parameters for connecting pipes in between magnets
+  for( uint pipeN = 0; pipeN < names.size(); pipeN++ ) {
+
+    if( lengths[pipeN] > 0 ) { continue; }
+    if( int(pipeN-1) < 0 ) { continue; }
+
+    double x = ( xCenters[pipeN-1] - lengths[pipeN-1]/2.*sin(thetas[pipeN-1]) + xCenters[pipeN+1] + lengths[pipeN+1]/2.*sin(thetas[pipeN+1]) ) / 2.;
+    double z = ( zCenters[pipeN-1] - lengths[pipeN-1]/2.*cos(thetas[pipeN-1]) + zCenters[pipeN+1] + lengths[pipeN+1]/2.*cos(thetas[pipeN+1]) ) / 2.;
+    double deltaX = (xCenters[pipeN-1] - lengths[pipeN-1]/2.*sin(thetas[pipeN-1])) - (xCenters[pipeN+1] + lengths[pipeN+1]/2.*sin(thetas[pipeN+1]));
+    double deltaZ = (zCenters[pipeN-1] - lengths[pipeN-1]/2.*cos(thetas[pipeN-1])) - (zCenters[pipeN+1] + lengths[pipeN+1]/2.*cos(thetas[pipeN+1]));
+    double l = sqrt( pow(deltaX, 2) + pow(deltaZ, 2) );
+
+    // Small air gap between connecting and magnet beam pipes to avoid G4 overlap errors
+    l -= 0.5;
+
+    double theta = atan( deltaX / deltaZ );
+
+    xCenters[pipeN] = x;
+    zCenters[pipeN] = z;
+    lengths[pipeN] = l;
+    thetas[pipeN] = theta;
+    rOuters1[pipeN] = rOuters2[pipeN-1];
+    rOuters2[pipeN] = rOuters1[pipeN+1];
+  }
+
+  // Add all pipes to the assembly
+  for( uint pipeN = 0; pipeN < xCenters.size(); pipeN++ ) {
+
+    ConeSegment s_tube( lengths[pipeN] / 2.0, rOuters1[pipeN] - thickness, rOuters1[pipeN], rOuters2[pipeN] - thickness, rOuters2[pipeN] );
+    ConeSegment s_vacuum( lengths[pipeN] / 2.0, 0, rOuters1[pipeN] - thickness, 0, rOuters2[pipeN] - thickness );
+
+    Volume v_tube("v_tube_" + names[pipeN], s_tube, m_Al);
+    Volume v_vacuum("v_vacuum_" + names[pipeN], s_vacuum, m_Vacuum);
+
+    v_tube.setVisAttributes(description.visAttributes( vis_name ) );
+
+    assembly.placeVolume(v_tube, Transform3D( RotationY(thetas[pipeN]), Position(xCenters[pipeN], 0, zCenters[pipeN])));
+    assembly.placeVolume(v_vacuum, Transform3D( RotationY(thetas[pipeN]), Position(xCenters[pipeN], 0, zCenters[pipeN])));
+
+  }
+
+  // Final placement
+  auto pv_assembly =
+    description.pickMotherVolume(sdet).placeVolume( assembly, Position(0.0, 0.0, 0.0));
+
+  sdet.setPlacement(pv_assembly);
+
+  assembly->GetShape()->ComputeBBox();
+
+  return sdet;
+}
+
+DECLARE_DETELEMENT(FarBackwardsBeamPipes, create_detector)
```

Comment:
Regarding the naming, we already have:
https://github.com/eic/epic/blob/c097fe24f3ba4b6ab1a549728c9919adbab04482/src/BackwardsBeamPipe_geo.cpp#L76
https://github.com/eic/epic/blob/c097fe24f3ba4b6ab1a549728c9919adbab04482/src/BackwardsCollimator.cpp#L80
https://github.com/eic/epic/blob/c097fe24f3ba4b6ab1a549728c9919adbab04482/src/BackwardsLumiVac_geo.cpp#L84
https://github.com/eic/epic/blob/c097fe24f3ba4b6ab1a549728c9919adbab04482/src/BackwardsTaggers_geo.cpp#L347

I suppose, we should rename those to have "Far" prefix as well?

Reply:
Well, better not change all those other file names.  I can just change this file name.